docs(core): update getting started pages (#31544)
nrwl/nx: docs/shared/getting-started/intro.md

---
title: Intro to Nx

> description: Nx is a powerful build system that streamlines development workflows, reduces errors, and dramatically cuts waste in projects of any scale.

Reviewer: should this include our newer AI based tag lines in anyway ?

> An AI-first build platform that connects everything from your editor to CI.
Helping you deliver fast, without breaking things.

---

# Introduction

Welcome to the official Nx documentation!

## What is Nx?

> Nx is a powerful, open source, technology-agnostic build platform designed to efficiently manage codebases of any scale. From small single projects to large enterprise monorepos, Nx is designed to **streamline your development workflows**, **reduce errors** and **dramatically reduce waste** by [saving engineering time](https://youtu.be/2BLqiNnBPuU) and cutting compute costs.

Reviewer: I didn't like having `reduce` in there twice. :) 
```suggestion
Nx is a powerful, open source, technology-agnostic build platform designed to efficiently manage codebases of any scale. From small single projects to large enterprise monorepos, Nx is designed to **streamline your development workflows**, **minimize errors** and **dramatically reduce waste** by [saving engineering time](https://youtu.be/2BLqiNnBPuU) and cutting compute costs.
```

{% youtube
src="https://www.youtube.com/embed/-_4WMl-Fn0w"
title="What is Nx?"
width="100%" /%}

## How Nx solves scaling

> Monorepos offer significant advantages. They reduce integration overhead between different software components, enabling faster cross-project changes, and improving code sharing and collaboration among teams. However, without proper tooling, as projects multiply, build times increase and management complexity grows, potentially negating the original benefits.

Reviewer: ```suggestion
Monorepos offer significant advantages. They reduce integration overhead between different software components, enabling faster cross-project changes, and improving code sharing and collaboration among teams. However, when projects multiply, build times increase and management complexity grows, potentially negating the original benefits.
```

No need to clarify without proper tooling due to you saying in the next sentence Nx solves this problem. IMO.

> Nx solves this by **providing solutions along the way for each step**, whether it's [for local development](/features), [integrating with your editor](/getting-started/editor-setup), [making your AI assistant smarter](/getting-started/ai-integration) and [scaling CI runs](/ci/features).

Reviewer: ```suggestion
Nx solves this by **providing solutions at each step** of your journey, whether it's [for local development](/features), [integrating with your editor](/getting-started/editor-setup), [making your AI assistant smarter](/getting-started/ai-integration) and [scaling CI runs](/ci/features).
```

## Start small, extend as you grow

Nx is built in a modular fashion, allowing you to adopt as little or as much as you'd like at any moment in your development lifecycle. You can **start with just the core and add additional capabilities incrementally** as your needs grow and complexity increases.

> **At the foundation is Nx Core**: a Rust-based, technology-agnostic task runner that builds a knowledge graph of your workspace, understanding project relationships and dependencies. This enables highly optimized and fast task execution regardless of technology stack. It runs `package.json` scripts in [TypeScript monorepos](/technologies/typescript/introduction) or Gradle tasks in [Java projects](/technologies/java/introduction).

Reviewer: ```suggestion
At the foundation is Nx Core, a Rust-based, technology-agnostic task runner. Nx Core creates a knowledge graph of your workspace, understanding project relationships and dependencies. This enables highly optimized and fast task execution regardless of technology stack. It runs `package.json` scripts in [TypeScript monorepos](/technologies/typescript/introduction) or Gradle tasks in [Java projects](/technologies/java/introduction) or can be extended to meet your projects specific needs
```

slight rewording. idk just personal preferences for this one.

> {% callout type="deepdive" title="Running NPM Scripts" %}

Reviewer: I'm really split on calling out the deep dive in the intro page. I think it's fine for now. but overall I want to make things as simple and straight forward on these initial pages and this feels almost like a diversion into people trying to grasp Nx off the bat.

You can get started with Nx by simply adding it to your existing project. This is all you need to begin leveraging Nx's core functionality.

For example, if you have a project with this `package.json`:

```json
{
  "name": "my-project",
  "scripts": {
    "build": "tsc",
    "test": "jest"
  }
}
```

Simply add Nx to your root `package.json`:

```json
{
  "devDependencies": {
    "nx": "latest"
  }
}
```

Now you can run:

```shell
nx build my-project
```

This will execute the `build` script from `my-project`'s `package.json`, equivalent to running `npm run build` in that project directory.

From there, you can gradually enhance your setup by adding features like [task caching](/features/cache-task-results), [task distribution](/ci/features/distribute-task-execution), and other powerful capabilities as your needs grow.

{% /callout %}

Nx Core works great alone, but you can incrementally add capabilities as needed:

> - [**Nx Cloud**](/ci): Scales builds through [remote caching](/ci/features/remote-cache), [distributed task execution](/ci/features/distribute-task-execution), [automated task splitting](/ci/features/split-e2e-tasks), and intelligent flaky task detection. Ensures CI performance scales linearly with codebase growth.

Reviewer: slight nit, task splitting is Nx Core feature, flaky task detection is a cloud thing. though work hand in hand. but more importantly why don't we have a flaky task detection page 😅. on my TODO now.

> - [**Nx Cloud**](/ci): Scales builds through [remote caching](/ci/features/remote-cache), [distributed task execution](/ci/features/distribute-task-execution), [automated task splitting](/ci/features/split-e2e-tasks), and intelligent flaky task detection. Ensures CI performance scales linearly with codebase growth.

Reviewer: ```suggestion
- [**Nx Cloud**](/ci): Scales builds through [remote caching](/ci/features/remote-cache), [distributed task execution](/ci/features/distribute-task-execution), [automated task splitting](/ci/features/split-e2e-tasks), and intelligent flaky task detection. Ensures CI performance scales with codebase growth.
```

idk if we should claim linearity

Author: no that was a mistake that got thrown in by AI when updating the sentence 😅. Linear scaling is what happens without Nx haha. Already fixed that locally on my branch

- [**Nx Console**](/getting-started/editor-setup): An editor extension that integrates Nx seamlessly into your development workflow with project details views, visual migration UI, command palette integration, and CI pipeline notifications. This provides enhanced developer experience directly within your editor.

- [**Nx AI**](/features/enhance-AI): Built-in LLM enhancement via Model Context Protocol (MCP) that gives AI assistants workspace architecture understanding, project relationships, and Nx documentation context. Transforms generic AI into architecturally-aware collaborators.

> - [**Nx Plugins**](/technologies): Technology-specific plugins automate scaffolding, code generation, and framework tasks for React, Angular, Node.js, and more. There's also a [Nx Devkit](/extending-nx/intro/getting-started) which enables you to build your own custom capabilities tailored to your organization's needs.

Reviewer: ```suggestion
- [**Nx Plugins**](/technologies): Technology-specific plugins automate scaffolding, code generation, and framework tasks for React, Angular, Node.js, and more. Extend Nx easily with [Nx Devkit](/extending-nx/intro/getting-started) to build custom capabilities tailored to your organization's needs.
```
